### ISSUE TITLE
Weird Decimal Round-off bug: Number with .43 as its decimal

### ISSUE THREAD

**Issue by yogeshbabu** [prisma/prisma#20881]
### Bug description

Field Type: Decimal
xx.43 becomes xx..43000000000001 ( only happens on .43 round-off, no other number as of now )
starts from 64 and goes till 99. In all other situations and other numbers (that i tested) seems to be fine.

example : 97.43 becomes 98.43000000000001

![image](https://github.com/prisma/prisma/assets/17112730/644baaee-aa23-49f4-8567-baadc601004d)
![image](https://github.com/prisma/prisma/assets/17112730/34bc60e2-5226-4cb8-adad-f74b35179294)
![image](https://github.com/prisma/prisma/assets/17112730/96f5c0f0-8630-4e91-b062-ff123b33d9ec)
![image](https://github.com/prisma/prisma/assets/17112730/00a64d08-20d7-4d74-a472-12231012e7b8)

### How to reproduce

Simply create a decimal filed and using Prisma Studio or a simple form try to submit the value xx.43 ( 64 - 99 )

### Expected behavior

If i enter 65.43 it has to be 65.34 not 65.43000000000001

### Prisma information

```prisma
// Add your schema.prisma
```

```ts
// Add your code using Prisma Client
model product {
id String   @id @default(uuid())
product_name String
amountValue  Decimal
}
```

### Environment & setup

- OS: 
- Database: 
- Node.js version: 
OS: Win 10 & 11 , Ubuntu
DB : SQLite
Node.js : v18.16.1

### Prisma Version

```
prisma : 5.2.0
@prisma/client : 5.2.0
```

**Comment by SevInf** [comment]
Can confirm with the following schema and code:

```prisma
datasource db {
  provider = "sqlite"
  url      = "file:dev.db"
}

generator client {
  provider = "prisma-client-js"
  output   = "../node_modules/.prisma/client"
}

model product {
  id          String  @id @default(uuid())
  amountValue Decimal
}
```

```ts
const prisma = new PrismaClient()

const result = await prisma.product.create({
  data: {
    amountValue: new Prisma.Decimal('97.43'),
  },
})
console.log(result)
```

However, it might not be actually a bug within Prisma, but `NUMERIC` type behaviour in sqlite: they are not arbitrary precision, they are floating-point. Issue does not happen with any other provider.

**Comment by SevInf** [comment]
Related #20635

**Comment by imevanc** [comment]
> Related #20635

@SevInf - Could this be happening due to the use of the [decimal.js](https://github.com/MikeMcl/decimal.js#readme)

**Comment by SevInf** [comment]
@imevanc I don't think so:
- `Decimal.js` is used by all providers, but this issue happens only with sqlite
- Issue does not happen with vanilla `Decimal.js` without the roundtrip through  DB: `Decimal('97.43')` stays `97.43`
- Unlike other providers, sqlite's `NUMERIC`/`DECIMAL` is known not to be floating point rather than fixed point

**Comment by jessehansen** [comment]
We are having this same issue with a different value (`83261.85`) in postresql. I'm working on a minimal reproduction now...

**Comment by jessehansen** [comment]
Here's a small reproduction:
```
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = "postgresql://local:local@localhost:5434/test_things?schema=public"
}

model ThingWithDecimal {
  id     Int     @id @default(autoincrement())
  amount Decimal
}
```

```
async function main() {
  const amount = 83261.85
  console.log(`Thing amount: ${amount}`)
  const thing = await prisma.thingWithDecimal.create({
    data: {
      amount,
    },
  })
  console.log(`Created thing: ${thing.amount}`)
  const thingRetrieved = await prisma.thingWithDecimal.findUnique({
    where: { id: thing.id },
  })
  console.log(`Retrieved thing: ${thingRetrieved?.amount}`)
}
```
Output:
```
Thing amount: 83261.85
Created thing: 83261.85000000001
Retrieved thing: 83261.85000000001
```

I was able to work around the issue by always sending `amount: new Prisma.Decimal(amount)` to the `create` call.

**Comment by rstreefland** [comment]
This doesn't seem to be SQLite specific. I've run into the same issue with Postgres.

Inserting a row with `8.39` as a number, results in Prisma inserting `8.390000000000001` into the database. The problem goes away if I `new Decimal(8.39)` _before_ inserting it into the database. 

I'm using version `5.5.2` of Prisma and my column has the Decimal type.

**Comment by SevInf** [comment]
@jessehansen if you want to preserve the precision you should use `Decimal` - js `numberts` are floating-point and imprecise by definition. Issue OP is talking about is Sqlite-specific and happens with `Decimal` too.

**Comment by jessehansen** [comment]
@SevInf 
> Issue OP is talking about is Sqlite-specific and happens with `Decimal` too.

This does happen in Prisma Studio with the postgres backend - it's not just sqllite. To reproduce, just use my reproduction schema from above, run prisma studio, & try to insert `83261.85` as a value.

> js numbers are floating-point and imprecise by definition

I understand javascript numbers are prone to these types of errors, but this is something other than just js number imprecision - while running the reproduction you can see that the precise value is correct in JS. It's only when the value is persisted in the database that the imprecision comes in. 

Here's some relevant debug log output:

```
prisma:client prisma.thingWithDecimal.create({
  data: {
    amount: 83261.85
  }
}) +0ms
prisma:client Generated request: +0ms
prisma:client {
  "modelName": "ThingWithDecimal",
  "action": "createOne",
  "query": {
    "arguments": {
      "data": {
        "amount": 83261.85
      }
    },
    "selection": {
      "$composites": true,
      "$scalars": true
    }
  }
}
```

**Comment by janpio** [comment]
The difference is that you are writing a JS `number` value into a database `Decimal` field and get back a different value than what you put in (which is pretty much expected as the `83261.85` under the hood already is that different value), but this issue is about writing a `Prisma.Decimal` into a database `Decimal` field and then still getting incorrect data back.

**Comment by janpio** [comment]
Yes, but exactly that makes it a different issue from this issue #20881 here. This one is _specifically_ about the problem that `Prisma.Decimal` also does not help, which is pretty surprising.

For the Prisma Studio specific problem, the best spot to report this is probably https://github.com/prisma/studio/issues/. I am not sure if someone has reported it before. (The fix will probably be to automatically turn the number from the input field via `Prisma.Decimal` into a Decimal before sending it to the database, so it can be stored correctly.)

**Comment by bdetry** [comment]
I had the same issue with Prisma **5.13.0**. 

**Schema :**  

`overtimeBalance Decimal      @db.Decimal(6, 2)`

**Postgre Database :** 

<img width="95" alt="Image" src="https://github.com/user-attachments/assets/70db37f1-129f-4054-8bec-8fb48752d7d5" />



The output of the query was 3.76 instead of 3.75

I upgraded Prisma to **6.4**, cleared caches and restarted everything. **It's working fine now.**
